lookout/core/bytes_to_unicode_converter.py

```diff
@@ -0,0 +1,128 @@
+        self._content_str = content.decode(errors="replace")
+        self._lines = self._content_str.splitlines(keepends=True)
+        self._byte_to_str_offset = self._build_bytes_to_str_offset_mapping(content)
+        self._lines_offset = self._build_lines_offset_mapping(self._content_str)
+
+    def convert_content(self):
+        """Convert byte content (or code) to unicode."""
+        return self._content_str
+
+    def convert_uast(self, uast: bblfsh.Node) -> bblfsh.Node:
+        """
+        Convert uast Nodes bytes position to unicode position.
+
+        UAST is expected to correspond to provided content.
+        :param uast: corresponding UAST.
+        :return: UAST with unicode positions.
+        """
+        uast = bblfsh.Node.FromString(uast.SerializeToString())
+        if not self._content:
+            return uast
+        for node in self._traverse_uast(uast):
+            for position in (node.start_position, node.end_position):
+                if position.offset == 0 and position.col == 0 and position.line == 0:
+                    continue
+                new_position = self._get_position(self._byte_to_str_offset[position.offset])
+                for attr in ("offset", "line", "col"):
+                    setattr(position, attr, getattr(new_position, attr))
+        return uast
+
+    def _get_position(self, offset: int) -> bblfsh.Position:
+        """Get new position for unicode string offset."""
+        line_num = numpy.argmax(self._lines_offset > offset) - 1
+        col = offset - self._lines_offset[line_num]
+        line = self._lines[line_num]
+        if len(line) == col:
+            if line.splitlines()[0] != line:
+                # ends with newline
+                line_num += 1
+                col = 0
+        return bblfsh.Position(offset=offset, line=line_num + 1, col=col + 1)
+
+    @staticmethod
+    def _build_lines_offset_mapping(content: str) -> numpy.ndarray:
+        if not content:
+            return numpy.empty(shape=(0, 0))
+        line_start_offsets = [0]
+        for d in content.splitlines(keepends=True):
+            line_start_offsets.append(line_start_offsets[-1] + len(d))
+        line_start_offsets[-1] += 1
+        return numpy.array(line_start_offsets)
+
+    @staticmethod
+    def _build_bytes_to_str_offset_mapping(content: bytes) -> Dict[int, int]:
+        """
+        Create a dictionary with bytes offset to unicode string offset mapping.
+
+        :param content: Bytes object which is used to create offsets mapping.
+        :return: Dictionary with bytes offset to unicode string offset mapping.
+        """
+        byte_to_str_offset = {0: 0}
+        byte_len_before = 0
+        content_str = content.decode(errors="replace")
+        for i, char in enumerate(content_str):
+            if char != "\ufffd":  # replacement character
+                byte_len_before += len(char.encode())
+            else:
+                byte_len_before += 1
```

Comment:
well, I test it and it creates one char per one char:
![image](https://user-images.githubusercontent.com/516045/54622021-d1a10600-4a79-11e9-8e60-24da929e018f.png)
 
(cannot attach notebook because viewer cannot handle it properly because of bad Unicode I suppose :) )

Do you have any example in the mind? I am not a pro in the unicode.

Reply:
Yes, looks like it is smart enough to replace by byte. OK.